policy: Pre-parse cached selectors, index by namespace

jrajahalme opened the pull request:
Parse selector requirement keys into labels when inserting into selector cache. Previously the labels of each existing cached selector (and `netip.Prefix` for a CIDR selector) were re-parsed and re-allocated on each match operation, used when new selectors or identities are added to the selector cache. This was essentially an O(n^2) operation on policy ingestion and identity updates.

Index cached selectors and identities by namespace so that it is possible to avoid futile matches rather than making the match operation itself fail early for a non-matching namespace. This speeds up namespaced policy ingestion and identity updates.

These changes make CIDR policy resolution in the new BenchmarkResolveCIDRPolicyRules ~4x faster, using 90% less memory and 96% less allocations.

jrajahalme committed:
policy: Pre-parse fqdnSelector

Store the fqdn identity label into `fqdnSelector` so that it does not
need to be re-created on every match operation.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/labels/array.go:
```diff
@@ -168,7 +172,11 @@ func (ls LabelArray) Has(key string) bool {
+	return ls.GetLabel(keyLabel)
+}
+
+func (ls LabelArray) GetLabel(keyLabel Label) string {
```
Should this take an explicit source? That might cut down on bugs.

jrajahalme replied:
IMO not, as the source is part of the `Label` passed in.

squeed commented on pkg/policy/selectorcache_selector.go:
```diff
@@ -104,23 +108,127 @@ func (f *fqdnSelector) remove(dnsProxy identityNotifier) {
+// Matches returns true if the Requirement matches the input Labels.
+// This is structurally the same as k8sLbls.Requirement.Matches(),
+// but takes LabelArray as an argument instead of a string.
+func (r *requirement) Matches(logger *slog.Logger, ls labels.LabelArray) bool {
```
If we're going to implement our own Matches, then we can get rid of dot-delineated labels entirely! They only existed to get labels with colons past the k8s parser.

squeed replied:
Can we move this to a separate package, and maybe copy some of the k8s unit tests?

jrajahalme replied:
Picking a separate package has some constraints due to dependencies:
- operates on types from `pkg/labels`
- converts from types in `pkg/policy/api` and `pkg/k8s/slim/k8s/apis/selection`

Maybe a sub-package in `pkg/policy` would work best?

jrajahalme replied:
Added `pkg/k8s/selector` for this, as there in fact was no dependency on `pkg/policy/api` after all.

We do not have unit tests for `pkg/k8s/slim/k8s/apis/labels` where this is adapted from, so did not add any here for now.

squeed commented on pkg/policy/selectorcache.go:
```diff
@@ -29,27 +29,51 @@ type scIdentity struct {
+	id := &scIdentity{
+		NID:       nid,
+		lbls:      lbls,
+		namespace: lbls.Get(labels.LabelSourceK8sKeyPrefix + k8sConst.PodNamespaceLabel),
```
namespace is a useful predicate to push, but I could see us adding additional ones. Such as the presence of FQDN or CIDR sources?

jrajahalme replied:
yes, and similar (but different) for selectors.

squeed commented on pkg/policy/selectorcache.go:
```diff
@@ -432,6 +510,23 @@ func (sc *SelectorCache) removeSelectorLocked(selector types.CachedSelector, use
+
+			namespaces := sel.source.selectedNamespaces()
+			if len(namespaces) == 0 {
+				// use empty namespace string for selectors without namespace
```
There are two cases here: when the selector selects no namespaces (the "" case), and when the selector selects an unknown set of namespaces (e.g. NOT IN). We need to handle these differently.

squeed replied:
on second thought, NOT IN selectors are quite uncommon, it's probably OK to merge them.

jrajahalme replied:
We may have a bug relating to NOT IN for namespaces, but I've tried to retain the current behavior, just filter earlier via an index.

More specifically, if a selector has both IN A and NOT IN B, then the full match will be skipped for identities in any namespace other than A, and then the full match will not match A due to NOT IN B.

One way to fix this is to detect the presence of a namespace with other than IN requirement and then treat the selector as having no namespace requirements, leading to full match against all identities.

jrajahalme replied:
Added a new commit to fix this bug, no tests for it yet, though.

squeed replied:
Whew, selectors are AND'ed together. So I agree, this is unlikely to be a bug.

squeed commented on pkg/policy/selectorcache.go:
```diff
@@ -512,6 +607,10 @@ func (sc *SelectorCache) CanSkipUpdate(added, deleted identity.IdentityMap) bool
 func (sc *SelectorCache) UpdateIdentities(added, deleted identity.IdentityMap, wg *sync.WaitGroup) (mutated bool) {
+	// set of namespaces to scan for updates. Always scan selectors that have no namespace
+	// requirements.
+	namespaces := map[string]struct{}{"": {}}
```
tracking the set of namespaces seems awkward. Can we instead shortcut in the `for numericID := range added {` section instead? That has the advantage of skipping an unnecessary call to `matches()` in the case where the identity's namespace cannot be selected by the selector.

jrajahalme replied:
We should already avoid most of the unnecessary `matches()` calls. Note that selectors without any namespace requirements may match identities in any namespaces, so all added identities must be matched against selectors without namespace requirements (indexed at "" namespace).

The case where unnecessary matching happens if we have multiple added identities in different namespaces on the same call, where we would try to match them against namespaced selectors in both namespaces.

We could clean this up by:
- handling deletes and adds separately
- track the namespace of each (added) identity so that only identities in a given namespace are matched against selectors that select that same namespace

But we still want to scan through each selector at most once, e.g., not scan though selectors separately for each added identity.

jrajahalme replied:
Added a new commit implementing the 2nd bullet above.

jrajahalme replied:
squashed that in.

jrajahalme committed:
policy: Fix selector namespace filtering

Consider selector possibly selecting any namespace if the selector
requirement has any operator than "In", "Equals", or "DoubleEquals". For
example, if the requirement is "NotIn", then any namespace *not in* the
values would be selected. In this kind of cases we skip namespace
filtering and perform a full match.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy/api: Parse CIDRGroupSelector as an EndpointSelector

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

mhofstetter reviewed (approved):
`envoy` :heavy_check_mark:

derailed commented on pkg/k8s/network_policy_test.go:
```diff
@@ -497,8 +498,9 @@ func TestParseNetworkPolicyNoSelectors(t *testing.T) {
 	}}
-	l3 := policytypes.ToPeerSelectorSlice(cidrRuleSlice.GetAsEndpointSelectors())
-	l3 = append(l3, policytypes.ToPeerSelectorSlice(cidrRuleSlice)...)
+	//l3 := policytypes.ToPeerSelectorSlice(cidrRuleSlice.GetAsEndpointSelectors())
```
nit delete

jrajahalme replied:
thanks, will do!

jrajahalme replied:
deleted :-)

derailed commented on pkg/labels/array.go:
```diff
@@ -168,9 +179,13 @@ func (ls LabelArray) Has(key string) bool {
 // ["cidr:1.1.1.1/32"].Has("cidr.1.0.0.0/8") => true
```
nit: Has->Get

jrajahalme replied:
This is actually a bit more intricate! CIDR labels do not have any value, so `Get` will return the same empty string regardless if the key matches any label in the array. Will fix the comment.

jrajahalme replied:
changed the comment.

derailed commented on pkg/labels/set.go:
```diff
@@ -0,0 +1,39 @@
+// Map keys are treated as Label with "k8s" source.
+type K8sSet map[string]string
+
+func (s K8sSet) HasLabel(l *Label) (exists bool) {
```
Seems we could `DRY` these up by leveraging `LookupLabel` to check the label or pull the value.
If so same applies to `Labels` above

jrajahalme replied:
Will consider, upstream matching logic I translated from used both `Has` and `Lookup`, but not `Get`.

jrajahalme replied:
done!

derailed commented on pkg/labels/array.go:
```diff
@@ -103,11 +103,18 @@ nextLabel:
 // ["k8s:foo=bar"].Intersects(["any:foo=bar"]) == true
 // ["any:foo=bar"].Intersects(["k8s:foo=bar"]) == false
 func (ls LabelArray) Intersects(needed LabelArray) bool {
+	for _, n := range needed {
```
nit: use index vs value range?

jrajahalme replied:
will change to index. Noticed that in these function even small overheads add up fast when adding new selectors to an already large set.

jrajahalme replied:
changed this and some other places to use the index instead. Benchmark showed ~15% speed up!

derailed commented on pkg/labels/array.go:
```diff
@@ -103,11 +103,18 @@ nextLabel:
+}
+
+func (ls LabelArray) IntersectsLabel(lbl Label) bool {
 	for _, l := range ls {
```
nit: same as above

jrajahalme replied:
will fix!

jrajahalme replied:
done!

derailed commented on pkg/labels/array.go:
```diff
@@ -168,9 +179,13 @@ func (ls LabelArray) Has(key string) bool {
 // ["cidr:1.0.0.0/8"].Has("cidr.1.1.1.1/32") => false
 func (ls LabelArray) Get(key string) string {
-	keyLabel := parseSelectLabel(key, '.')
+	keyLabel := ParseSelectDotLabel(key)
```
Seems we could `dry` this up and leverage lookupLabel?

jrajahalme replied:
`Get` is an implementation of an upstream k8s interface, but yes, the implementation can be changed :-)

jrajahalme replied:
Changed the LabelMatcher interface to only have LookupLabel method, and using it also internally for all the applicable existing methods.

joamaki commented on pkg/policy/selectorcache.go:
```diff
@@ -86,27 +171,40 @@ type SelectorCache struct {
 
-// GetVersionHandleFunc calls the given function with a versioned.VersionHandle for the
+func (sc *SelectorCache) toReadTxn(rtxn statedb.ReadTxn) SelectorReadTxn {
+	return SelectorReadTxn{Txn: rtxn, Rev: sc.selections.Revision(rtxn)}
```
Holding onto a ReadTxn can be problematic since it holds onto all the data in the snapshot. We can fairly easily change `ReadTxn` to take an optional set of tables to narrow down the snapshot to improve this.

Do we need to hold onto to the whole selections table or do we just parts of it? We could also have a new query which returns e.g. a prefix of a table to which new queries can be made.

jrajahalme replied:
We have two reasons for holding on to the ReadTxn:
1. Envoy policy update after a full policy recomputation needs to be built from the same version as the bpf maps. 
2. Envoy policy update after an incremental policy update needs to be built from the same version as the bpf map updates.

In both cases the ReadTxn is closed (`nil`led) soon after via `EndpointPolicy.Ready()` .

Each endpoint does this independently, so there are multiple ReadTxn's held during policy updates, possibly to different versions, depending on the timing of the endpoint regeneration vs. selector churn. Holding on to a smaller section of the db would help in the general case, but in this minimal instance we have a separate db just for this purpose, so as of now there would be no difference.

joamaki replied:
Thanks for clarifying. This seems alright then!

joamaki commented on pkg/policy/selectorcache_selector.go:
```diff
@@ -4,24 +4,64 @@
+type Selectors = types.Selectors
+type SelectorReadTxn = types.SelectorReadTxn
+
+type selections struct {
```
Any particular reason why this is unexported? Do we not have cases where we'd want to query this table from outside this package? 

Also please either change the fields to be exported or implement `MarshalJSON` so that this table can be included in the sysdump (the whole DB is dumped as JSON in sysdump).

jrajahalme replied:
No, will export this.

jrajahalme replied:
exported!

joamaki commented on pkg/policy/selectorcache_selector.go:
```diff
@@ -4,24 +4,64 @@
+	},
+
+	FromKey: index.String,
+
```
Add `FromString: index.FromString` so one can do e.g. `db/get selections <key>` etc.

jrajahalme replied:
will do, thanks for the tip!

jrajahalme replied:
Done.

joamaki commented on pkg/policy/selectorcache_selector.go:
```diff
@@ -4,24 +4,64 @@
+}
+
+// SelectionsIndex is the primary index for selections indexing the key of the cached selector
+var SelectionsIndex = statedb.Index[selections, string]{
```
Since `selections` is not exported it doesn't really make sense to export this either

jrajahalme replied:
Exporting `selections` soon.

jrajahalme replied:
Done.

derailed reviewed (commented):
@jrajahalme Great work!

jrajahalme committed:
policy: Pre-parse labelIdentitySelector

Use the cached string and requirements with pre-parsed labels instead of
api.EndpointSelector in labelIdentitySelector.

labels.LabelArray is changed to add HasLabel(), GetLabel() and
LookupLabel() functions that take a pre-parsed Label as an argument
instead of a string.

This makes CIDR policy resolution in BenchmarkResolveCIDRPolicyRules 70%
faster, with 90% less memory used and 95% less allocations.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Index identities by namespace

Add a namespace index to identities so that ones with non-matching
identities can be trivially skipped when adding or removing selectors
to/from the selector cache.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Index cached selectors by namespace

Index cached selectors by namespace so that ones with non-matching
namespaces can be trivially skipped when adding or deleting identities.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

odinuge commented on pkg/policy/selectorcache.go:
```diff
@@ -324,49 +398,166 @@ type identityNotifier interface {
+// UpdateIdentities(). Similar clean-up while scanning can not be done for the main selectors map as
+// there is no need to iterate over it otherwise.
+// This could probably be done more efficiently via a timer job periodically.
+func (sc *SelectorCache) finalizer(sel *identitySelector) {
```
Do you have any production testing of this? The benchmarks in the PR seems a bit theoretical for a single use-case, so very curious to see what issue this is addressing and how it affects cpu and memory usage. It seems a bit worrying that this will only happen periodically now, potentially resulting in a bigger selector cache that use more memory than previously.

odinuge replied:
I think the overall idea here is great, but I think it would be very very useful to have some real production-style workloads that actually show how this will affect end users before merging. This since a regression in this subsystem is (often) pretty hard to debug with the current set of metrics - and it can cause lot of bad side-affects and bad issues for end users if this does not work fully as expected.

And for real production-style workloads that would be something that stresses this code-path from all angles - not just one of them. So churn in actual policies, chrun in pods, churn in policy regenerations (for various reasons I guess - and v1.16 does loads of them), churn in identities etc.

marseel replied:
> Do you have any production testing of this? The benchmarks in the PR seems a bit theoretical for a single use-case, so very curious to see what issue this is addressing and how it affects cpu and memory usage. It seems a bit worrying that this will only happen periodically now, potentially resulting in a bigger selector cache that use more memory than previously.

> And for real production-style workloads that would be something that stresses this code-path from all angles - not just one of them. So churn in actual policies, chrun in pods, churn in policy regenerations (for various reasons I guess - and v1.16 does loads of them), churn in identities etc.

FYI, we are running scale test in CI that regularly exercises this type of scenario - churn of policies/identities/pods etc.
We are looking at various metrics like cpu/memory/policy implementation latency etc, but not the actual correctness.

Taking into account complexity in this PR, I think it's better to perform such testing before merging PR.
The other parts of your questions around correctness and debuggability , I will leave to reviewers and Jarno to answer

odinuge commented on pkg/policy/selectorcache.go:
```diff
@@ -324,49 +398,166 @@ type identityNotifier interface {
-	// Make the FQDN subsystem aware of this selector
-	sc.localIdentityNotifier.RegisterFQDNSelector(source.selector)
+	if len(sel.users) > 0 {
+		sc.logger.Warn(
```
Should this instead be an error since this means something in the policy or fqdn subsystem is behavior inconsistent? Also maybe add that it can/will lead to unexpected policy drops.

odinuge replied:
Or, if this is some other leak that keeps the endpoint objects in memory after an endpoint is deleted - that also seems like a memory leak we want to catch properly.

jrajahalme replied:
I've since got rid of the warning condition here, will post an update once I get some more clean-ups in.

odinuge commented on pkg/policy/selectorcache.go:
```diff
@@ -580,61 +715,70 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted identity.IdentityMap, w
+		for ns, nsAdded := range namespaces {
+			// Iterate through all locally used identity selectors and
+			// update the cached numeric identities as required.
+			for wp := range sc.selectorsByNamespace[ns] {
```
Would it be better to move this to a separate function that makes it clear to the reader it has this side-effect? It seems a bit convoluted to do this inside the `UpdateIdentities` code directly - and I feel like later refactors can result in correctness bugs if its not fully understood.

jrajahalme replied:
Yes.

odinuge commented on pkg/policy/selectorcache.go:
```diff
@@ -55,20 +141,20 @@ type userNotification struct {
 
 	// map key is the string representation of the selector being cached.
-	selectors map[string]*identitySelector
+	selectors map[string]weak.Pointer[identitySelector]
```
Would it be better to create a struct here that contains both the current data from `selectors` and `selectorsByNamespace`? You already did that for the `scIdentityCache` struct, and I think that will make the indexing code a bit easier to understand, test and free from correctness bugs.

jrajahalme replied:
Fair, will do :-)

jrajahalme replied:
Done!

odinuge commented on pkg/policy/selectorcache.go:
```diff
@@ -512,22 +641,28 @@ func (sc *SelectorCache) CanSkipUpdate(added, deleted identity.IdentityMap) bool
 func (sc *SelectorCache) UpdateIdentities(added, deleted identity.IdentityMap, wg *sync.WaitGroup) (mutated bool) {
+	// Map of namespaces to scan for updates with added identities in the map value. All
+	// identities are matched against selectors that have no namespace requirements.
+	namespaces := map[string]identity.NumericIdentitySlice{"": {}}
```
It feels a bit hard to read through the code when this defaults to `{"":{}}`, meaning it will _always_ loop through the non-namespaced value. For adds you populate this map with the actual nids, but for deletes that is not the case, and all of them are looped through.

Am I missing something, or could just just treat adds and deletes the same way, since it seems like the assumption that the nid of the last version of an identity is the only namespace we look in anyways.

jrajahalme replied:
The intent is to limit the set of cached selectors we need to scan through by skipping namespaced selectors when none of the affected identities (added or deleted) are in the to-be-skipped namespace. We must check all non-namespaced selectors as they could select identities in any namespace.

`namespaces` is updated for both `deleted` and `added`, but for deleted we only need the namespace, rather than the affected identities in that namespace.

Since (non-cachad) matching on selectors is expensive, we are scanning each (potentially) affected selector only once, and handled both deletes and adds within the selector loop. We could be more precise and check deletions only in namespaces that had deletions (in addition to no-namespace `""`), as currently we check for deletions also in namespaces that only had additions.

odinuge commented on pkg/policy/types/requirements.go:
```diff
@@ -0,0 +1,226 @@
+// for single-valued values.
+// +deepequal-gen=true
+type Requirement struct {
+	key      labels.Label       // pre-parsed Label
```
I found this pretty confusing to being with since the `labels.Label` object has both a key _and_ and value, together with the source. I think this can also cause a lot of confusion elsewhere as well, since passing around a `labels.Label` and treating it as key is pretty confusing. Because then the actual value _sometimes_ matters, but other times not. 

I eg. spent a decent amount of time understanding why the `LookupLabel(label *Label)` was expected to return a different `value` than the one provided as the `label` argument.

So I think its better here to either keep the actual key and source separate here, or alternative create a new struct that has the key _and_ the source in it.

jrajahalme replied:
`key` here is a `Label` since that is what the former string key was parsed to (previously) when matching an identity against a selector. You did not see that since it was hidden in the `labels` package :-) So this is functionally non-changing refactor.

But you are right that there is scope for further refactoring here.

odinuge reviewed (commented):
I have some initial questions and comments, but never fully got through the full diff here since its so large. Mainly questions around performance and how this is/has been tested in production style workloads (and not just small repeated benchmarks). We have over the past set of releases seen various performance related changes like this that have caused performance regressions, so I'm trying to fully understand how this can affect the policy and fqdn subsystem overall.

I think personally it would make sense to split some of this into separate changes to make it easier to review and reason about for end users. eg. I also think the sharding per-ns is great - and its something we have thought about implementing ourselves as well - but some other things are a bit harder to reason about here when it comes to performance and scaleability.

Thanks

jrajahalme committed:
policy: Add test for wildcard IPv4 CIDR selector

Add test to verify the wildcard IPv4 selector (0.0.0.0/0) selects both
world-ipv4 and any IPv4 CIDR identity.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy/api: Use world label in wildcard CIDR selectors

Use world label in CIDR selectors instead of a zero-length prefixes. Thus
a CIDR "0.0.0.0/0" selector will translate to a "reserved:world" selector
(in an IPv6 disabled configuration). This makes it possible to remove the
addition of world selectors when wildcard selectors are added.

Do not add extra selectors with "reserved:world" label requirement (or
dual-stack equivalents) when a CIDR selecting all IPv4 or IPv6 addresses
is ingested. Adding the extra world label selector will no longer expand
the set of selected identities, since the world selector has already been
used instead of a zero-length CIDR selector.

The removed logic was also partially based on string matching on forms
that would not match for selectors with a CIDR "1.1.1.1/0" for example,
even though the resulting CIDR selector would be the same
("cidr.0.0.0.0/0").

This change allows CIDR selector parsing to internal
representation to be simplified, as each CIDR selector will map to
exactly one internal EndpointSelector.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Remove Requirements from api.EndpointSelector

Keep pre-computed Requirements only in Selector implementations in
pkg/policy/types. Use these in places where EndpointSelector was used for
matching in production code previously.

This simplifies policy/api code, and keeps the faster matching in one
place (policy/types.Selector implementations). These same Selectors are
now also used internally in SelectorCache for updating the cached
selections.

Note that the matching interface of api.EndpointSelector has been moved
to unit test code. Due to dependency cycle avoidance the unit test
matching logic in policy/api can not use the new matching logic in
policy/types. For selector matching in production code, create a
types.LabelSelector from api.EndpointSelector and use that for
matching. Note that for this to work the api.EndpointSelector MUST HAVE
the key prefixed with a label source (e.g., "k8s."), otherwise any dot
separated prefix will be parsed as the label's source and removed from
the label key!

API selector types now all have a SelectorKey() method, used for getting
a key string before a new Selectors are created from them. This method
should be fast, and api.EndpointSelector keeps the internal
cachedLabelSelectorString member for this purpose. This cached string is
still updated after each modification to api.EndpointSelector.

PeerSelector interface used to be a "marker" interface with
IsPeerSelector() function that did nothing. Since we added another
function (SelectorKey()) to the interface, it is not a pure marker
anymore, and IsPeerSelector() function is no longer needed, so remove it.

Since also subject selectors can be cached in selector cache, rename
PeerSelector interface as APISelector, as it must be implemented by all
policy.api types that can be used as types.Selectors.

Stop hauling logger to selector matching functions, as they use validated
inputs and should never fail. To help avoid breaking this invariant, use
logging.DefaultSlogLogger as a "soft panic" to fail Cilium CI if these
are hit during CI runs.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: refactor scIdentityCache to a struct

Refactor scIdentityCache to a struct.

Keep scIdentities in a map via a pointer so that it is possible to point
to them via a secondary index to be added in the next commit.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
labels: Add LabelMatcher interface

Add a new LabelMatcher interface with a sole LookupLabel function that
uses a pre-parsed *Label as an argument. Implement it in Labels,
LabelArray, and in a new K8sSet, that can be used to cast bare sting maps
to a LabelMatcher.

Use iteration index into LabelArray rather than using a temporary
Label. This makes a diffence due to LookupLabel used in some inner loops
in policy selector matching.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

derailed commented on pkg/container/set/set.go:
```diff
@@ -168,6 +168,16 @@ func (s Set[T]) Equal(o Set[T]) bool {
 }
 
+// DeepEqual is same as Equal due to Set keys being comparable.
+func (s *Set[T]) DeepEqual(o *Set[T]) bool {
```
We are mixing semantics between value and pointer receiver.
Looking at the constructor we are pushing value semantics is this intended?

jrajahalme replied:
Some of the design goals/constraints for `set.Set[T]` were/are:
- Allow storing by value to avoid unnecessary memory dereference, especially when the included map is already accessed by reference
- It is small enough to be passed by value without performance degradation, nicer when passing as a parameter from a struct that stores it by value, no need to play with pointer and pointer dereference operators
- Modifying operations need the pointer receiver to be able to update the `single` member, and need to return a `changed bool` return value to let the caller know that the struct was changed for the caller to update any stored values when a temporary was modified (e.g., map values containing the set by value was changed during iteration via a temporary)
- `DeepEqual` requires a pointer receiver and a pointer parameter for the generated deepequal functions to be able to call it properly

So the mixing of value and pointer receivers is somewhat intentional, but it should be possible to change value receivers to pointer receivers without any problems. Not in this PR, though. Still the value semantics here is a bit convoluted due to design goal of allowing storing by value for efficiency.

derailed replied:
Thank you for the details Jarno!
I always get a bad oompa-loompa vibe when I see mixed semantics. 
Ultimately the question should be is this object meant to be shared or copied? regardless of the underlying mechanic

derailed reviewed (commented):
@jrajahalme Nice work!

odinuge reviewed (dismissed):
And to be explicit about the comment above, I'm effectively NACKing the weak pointer part that relies on the go GC to do selector cache cleanups.

I don't think its a good idea to rely on GC to "limit" the complexity of the hot-path in cilium - aka. to cleanup the entries. For a normal cache I think its fine (and great), but currently the selector cache is doing a lot of work, and if this ends up bloating it - it results in much higher cpu and memory usage, as well as contributing to lock contention. On identity updates, the selector cache is looping through this data structure to find matching selectors - and basing that list purely on GC to cleanup will cause side-effects in terms of performance. 

If entries stay there for longer, it will for sure have a (very) negative effect on our workloads, and I don't think the benefit is worth the trade-off here.

squeed commented on pkg/policy/api/cidr.go:
```diff
@@ -86,42 +76,15 @@ type CIDRSlice []CIDR
 // GetAsEndpointSelectors returns the provided CIDR slice as a slice of
 // endpoint selectors
 func (s CIDRSlice) GetAsEndpointSelectors() EndpointSelectorSlice {
-	// If multiple CIDRs representing reserved:world are in this CIDRSlice,
```
Why is it safe to remove this code? The `world` identity does not have `0.0.0.0/0` or `::/0`, meaning we have to hack it up here so that a zero-length prefix selects it.

jrajahalme replied:
This is safe due to a full wildcard being replaced with a `reserved:world` label selector (or qualified one for dual stack), which matches the `world` identity.

jrajahalme replied:
`reserved:world` label selector also matches all CIDR identities due to them having the `reserved:world` label in addition to the CIDR label.

squeed replied:
Aha, I see it now. Cute. You fixed it up in IPStringToLabel.

squeed commented on pkg/policy/api/selector.go:
```diff
@@ -355,9 +355,13 @@ func (n *EndpointSelector) IsWildcard() bool {
+	return n.SanitizeWithKeyExteder(labels.DefaultKeyExtender)
+}
+
+func (n *EndpointSelector) SanitizeWithKeyExteder(extender labels.KeyExtender) error {
```
typo?

jrajahalme replied:
Yes, but fixed in a later commit.

sayboras reviewed (approved):
Looks good for cilium/proxy related files

joamaki reviewed (approved):
sig-k8s & sig-lb LGTM

squeed commented on pkg/policy/types/requirements.go:
```diff
@@ -0,0 +1,264 @@
+// escape labels.LabelArray implementing the interface to the heap, causing large memory
+// overhead. When the compiler specializes the generic type to a concrete type (not an interface),
+// it can perform correct escape analysis and avoid unnecessary heap allocations.<
+func MatchesRequirement[T labels.LabelMatcher](logger *slog.Logger, r *Requirement, ls T) bool {
```
I'd suggest this return an error instead of taking a logger; callers can handle the error as they see fit.

Requirements are parsed from k8s label selectors, which are validated. So in practice I don't think this is possible, but the "chain" is long and seems fragile, so-to-speak.

jrajahalme replied:
Then I would need to decide on each call site what to do with the error, which would only happen if values were not properly validated.

IMO the "right thing" to do would be to panic, but we should never do that in production. The next best thing to do is to log an error that will fail our CI. If we always log the error then we should do it here instead of all the call sites.

We can get rid of passing the logger by using `logging.DefaulSlogLogger` instead. Since this is a "soft panic" and never expected in production, this would be acceptable, IMO.

jrajahalme replied:
@squeed Added a commit to address this, please take a look!

squeed replied:
thanks, that makes things cleaner.

squeed commented on pkg/policy/types/selector.go:
```diff
@@ -6,14 +6,527 @@ package types
+
+	// KeyString is a unique string for the given selector that is used as a key in
+	// selector maps. Must never be an empty string.
+	KeyString() string
```
This seems like an internal implementation detail that belongs in `pkg/policy/types`? I see a lot of coupling between the internals of `pkg/policy/api` and `pkg/policy/types/selector.go`, especially in translating the api to selectors.

Perhaps there should be even less logic in `pkg/policy/api`? Especially with the potential transition to a bifurcated selectorcache, we could maybe even get rid of the PeerSelector interface.

jrajahalme replied:
IMO caching the string representation in `api.EndpointSelector` (cachedString) is valuable as computing the string is expensive, especially for a LabelSelector.

PeerSelector was only used to mark API types that can function as peer Selectors in the policy, and as such it is valuable. Now that I added another method, it is not a pure "marker" interface any more. Maybe I should remove the marker method (`IsPeerSelector()`) and use a bit better name for `KeyString()` that carries the same meaning without any "peer" connotation? How about `SelectorKey() string`?

I tried using `String()` instead of `KeyString()` but there is some existing use of String() somewhere (maybe a cast to Stringer interface) that interfered with this plan.

jrajahalme replied:
Added a commit to address this, please have a second look!

tklauser commented on pkg/policy/k8s/service.go:
```diff
@@ -352,37 +352,32 @@ func serviceSelectorMatches(sel *api.K8sServiceSelectorNamespace, svc serviceDet
-		value = v.Value
-	}
-	return
+// Get implements labels.LabelMatcher; label source is ignored
```
nit:


```suggestion
// GetLabel implements labels.LabelMatcher; label source is ignored
```

tklauser commented on pkg/policy/k8s/service.go:
```diff
@@ -352,37 +352,32 @@ func serviceSelectorMatches(sel *api.K8sServiceSelectorNamespace, svc serviceDet
-// Has implements k8sLabels.Labels.
-func (l labelsMatcher) Has(label string) (exists bool) {
-	return labels.Labels(l).HasLabelWithKey(label)
+// Has implements labels.LabelMatcher.
```
```suggestion
// HasLabel implements labels.LabelMatcher.
```

tklauser commented on pkg/policy/k8s/service.go:
```diff
@@ -352,37 +352,32 @@ func serviceSelectorMatches(sel *api.K8sServiceSelectorNamespace, svc serviceDet
-// Lookup implements k8sLabels.Labels.
-func (l labelsMatcher) Lookup(label string) (value string, exists bool) {
-	v, ok := labels.Labels(l)[label]
+// Lookup implements labels.LabelMatcher
```
```suggestion
// LookupLabel implements labels.LabelMatcher
```
